cpp/include/rapidsmpf/shuffler/chunk.hpp

```diff
@@ -25,145 +27,335 @@ namespace rapidsmpf::shuffler::detail {
+        return chunk_id_;
+    }
+
+    /**
+     * @brief The number of messages in the chunk.
+     *
+     * @return The number of messages in the chunk.
+     */
+    constexpr size_t n_messages() const {
+        return n_messages_;
+    }
 
-    /// If not zero, the number of chunks of the partition expected to get from the
-    /// sending rank. Ignored when it is zero.
-    std::size_t const expected_num_chunks;
+    /**
+     * @brief Partition ID of the i-th message.
+     *
+     * @param i The index of the message.
+     * @return The ID of the partition.
+     */
+    inline PartID part_id(size_t i) const {
+        return part_ids_.at(i);
+    }
+
+    /**
+     * @brief The expected number of chunks of the i-th message.
+     *
+     * @param i The index of the message.
+     * @return The expected number of chunks for the message. Non-zero when the message
+     * is a control message, otherwise zero (data message).
+     */
+    inline size_t expected_num_chunks(size_t i) const {
+        return expected_num_chunks_.at(i);
+    }
 
-    /// If known, the size of the GPU data buffer (in bytes).
-    std::size_t const gpu_data_size;
+    /**
+     * @brief Whether the i-th message is a control message.
+     *
+     * @param i The index of the message.
+     * @return True if the message is a control message, false otherwise.
+     */
+    inline bool is_control_message(size_t i) const {
+        return expected_num_chunks(i) > 0;
+    }
 
-    /// Metadata of the packed `cudf::table` associated with this chunk.
-    std::unique_ptr<std::vector<uint8_t>> metadata;
+    /**
+     * @brief Get the data of the i-th message, as a new ChunkBatch.
+     *
+     * @param new_chunk_id The ID of the new chunk.
+     * @param i The index of the message.
+     * @param stream The CUDA stream.
+     * @return A new ChunkBatch containing the data of the i-th message.
+     * @note This will create a copy of the packed data. If there is only one message and
+     * the message is a data message, the buffers will be moved to the new ChunkBatch.
+     * Otherwise a new ChunkBatch will be created by copying data.
+     *
+     * @throws std::out_of_range if the index is out of bounds.
+     */
+    Chunk get_data(ChunkID new_chunk_id, size_t i, rmm::cuda_stream_view stream);
 
-    /// GPU data buffer of the packed `cudf::table` associated with this chunk.
-    std::unique_ptr<Buffer> gpu_data;
+    /**
+     * @brief Get the size of the metadata of the i-th message.
+     *
+     * @param i The index of the message.
+     * @return The size of the metadata of the message. Zero when the message is a
+     * control message, otherwise the size of `PackedData::metadata`.
+     */
+    inline uint32_t metadata_size(size_t i) const {
+        assert(i < n_messages());
+        assert(!meta_offsets_.empty());
+        assert(!is_control_message(i));
+        return i == 0 ? meta_offsets_[0] : meta_offsets_[i] - meta_offsets_[i - 1];
+    }
```

Comment:
I think it would ggod to be bit more defensive here 
```suggestion
    inline uint32_t metadata_size(size_t i) const {
        RAPIDSMPF_EXPECTS(i < n_messages());
        RAPIDSMPF_EXPECTS(!is_control_message(i));
        return i == 0 ? meta_offsets_.at(0) : meta_offsets_.at(i) - meta_offsets_.at(i - 1);
    }
```